packages/block-editor/src/components/story-update-progress/index.native.js

```diff
@@ -0,0 +1,293 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import { View } from 'react-native';
+
+/**
+ * WordPress dependencies
+ */
+import { Spinner } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import {
+	subscribeMediaUpload,
+	subscribeMediaSave,
+} from '@wordpress/react-native-bridge';
+
+/**
+ * Internal dependencies
+ */
+import styles from './styles.scss';
+
+export const STORY_UPLOAD_STATE_UPLOADING = 1;
+export const STORY_UPLOAD_STATE_SUCCEEDED = 2;
+export const STORY_UPLOAD_STATE_FAILED = 3;
+export const STORY_UPLOAD_STATE_RESET = 4;
+
+export const MEDIA_SAVE_STATE_SAVING = 5;
+export const MEDIA_SAVE_STATE_SUCCEEDED = 6;
+export const MEDIA_SAVE_STATE_FAILED = 7;
+export const MEDIA_SAVE_STATE_RESET = 8;
+export const STORY_SAVE_STATE_RESULT = 9;
+export const MEDIA_SAVE_MEDIAMODEL_CREATED = 10;
+
+export class StoryUpdateProgress extends React.Component {
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			progress: 0,
+			isSaveInProgress: false,
+			isSaveFailed: false,
+			isUploadInProgress: false,
+			isUploadFailed: false,
+		};
+
+		this.mediaUpload = this.mediaUpload.bind( this );
+		this.mediaSave = this.mediaSave.bind( this );
+	}
+
+	componentDidMount() {
+		this.addMediaUploadListener();
+		this.addMediaSaveListener();
+	}
+
+	componentWillUnmount() {
+		this.removeMediaUploadListener();
+		this.removeMediaSaveListener();
+	}
+
+	mediaIdContainedInMediaFiles( mediaId, mediaFiles ) {
+		if ( mediaId !== undefined && mediaFiles !== undefined ) {
+			for ( let i = 0; i < this.props.mediaFiles.length; i++ ) {
+				if ( mediaFiles[ i ].id === mediaId.toString() ) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	mediaUpload( payload ) {
+		const { mediaFiles } = this.props;
+
+		if (
+			this.mediaIdContainedInMediaFiles( payload.mediaId, mediaFiles ) ===
+			false
+		) {
+			return;
+		}
+
+		switch ( payload.state ) {
+			case STORY_UPLOAD_STATE_UPLOADING:
+				this.updateMediaUploadProgress( payload );
+				break;
+			case STORY_UPLOAD_STATE_SUCCEEDED:
+				this.finishMediaUploadWithSuccess( payload );
+				break;
+			case STORY_UPLOAD_STATE_FAILED:
+				this.finishMediaUploadWithFailure( payload );
+				break;
+			case STORY_UPLOAD_STATE_RESET:
+				this.mediaUploadStateReset( payload );
+				break;
+		}
+	}
+
+	mediaSave( payload ) {
+		const { mediaFiles } = this.props;
+
+		if (
+			this.mediaIdContainedInMediaFiles( payload.mediaId, mediaFiles ) ===
+			false
+		) {
+			return;
+		}
+
+		switch ( payload.state ) {
+			case MEDIA_SAVE_STATE_SAVING:
+				this.updateMediaSaveProgress( payload );
```

Comment:
I am wondering if we could utilize array destructuring for the `mediaSave` and `mediaUpload` functions to make it clear what is being passed from the payload to these event listeners and to ensure that we aren't passing extra data to some of the functions?